File: stack.cpp
#include <iostream>
using namespace std;

class Stack {
	struct node {
		int data;
		node *next;
	};

  private:
    node *head, *tail;

  public:
    Stack() {
      head = NULL;
      tail = NULL;
    }

  void push(int data) {
    if (!head) {
      head = new node;
			head -> data = data;
      head -> next = NULL;
			tail = new node;
      tail = head;
    } else {
			node *itr = new node;
      itr -> data = data;
			itr -> next = tail;
			tail = itr;
		}
  }

  void pop() {
    node *itr = new node;
  	if (head == tail) {
      itr = head;
      itr -> data = head -> data;
      head = NULL;
      tail = NULL;
			delete itr;
    } else {
      itr = tail;
      tail = tail -> next;
			cout << "Pop: " << itr -> data << endl << endl;
			delete itr;
    }
  }

  void display() {
		node *itr = new node;
		itr = tail;
		cout << "Nodes in stack from top to bottom: ";
		while (itr != NULL) {
			cout << itr -> data <<" ";
			itr = itr -> next;
		}
		cout << endl << endl;
	}
};

int main() {
  Stack stack;
  stack.push(1);
  stack.push(2);
  stack.push(3);
  stack.push(4);
	stack.display();

	stack.pop();
  stack.display();

	stack.pop();
	stack.display();

	stack.pop();
	stack.display();

  return 0;
}

File: stack.java
public class Main {
  public static void main(String[] args) {
    Stack stack = new Stack();
    stack.push(1);
    stack.push(2);
    stack.push(3);
    stack.push(4);
    stack.pop();
    stack.display();
  }
}

class Stack {
  class Node {
    int data;
    Node next;

    public Node(int data) {
      this.data = data;
      this.next = null;
    }
  }

  public Node head = null;
  public Node tail = null;

  public void push(int data) {
    Node node = new Node(data);
    if (head == null) {
      head = node;
      tail = head;
    } else {
      node.next = tail;
      tail = node;
    }
  }

  public Node pop() {
    Node itr = tail;
    if (tail == null) return null;
    else if (head == tail) {
      head = null;
      tail = null;
      return tail;
    } else {
      tail = tail.next;
      System.out.println("Popped: " + itr.data);
      return itr;
    }
  }

  public void display() {
    Node itr = tail;
    if (head == null) System.out.print("Empty stack");
    else {
      System.out.print("Nodes in stack: ");
      while (itr != null) {
        System.out.print(itr.data + " ");
        itr = itr.next;
      }
      System.out.println();
    }
  }
}
